### ISSUE TITLE
Confused about proper use of `add_step_closure`.  Why isn't it called *before* `xm.optimizer_step` in the examples?

### ISSUE THREAD

**Issue by hrbigelow** [pytorch/xla#5094]
## ❓ Questions and Help

Hi,

In the following example (abbreviated from [test_train_mp_mnist.py](https://github.com/pytorch/xla/blob/5c95ad2cba9c19f5db930d8f5102192bc33ad2c7/test/test_train_mp_mnist.py#LL165C1-L171C44)):

```python
  def train_loop_fn(loader, epoch):
    for step, (data, target) in enumerate(loader):
      optimizer.zero_grad()
      output = model(data)
      loss = loss_fn(output, target)
      loss.backward()
      xm.optimizer_step(optimizer)
      xm.add_step_closure(
          _train_update,
          args=(device, step, loss, tracker, epoch, writer),
          run_async=flags.async_closures)

```

Is this the correct order of things on the TPU side?

iteration 1:
- mark_step from loader
- loss computed
- optimizer step
  - mark_step
- closure 1 added
 
iteration 2:
- mark_step from loader
  - closure 1 run
- loss computed
- optimizer step
  - mark_step
- closure 2 added
...

It seems confusing that closure 1 is run during iteration 2 rather than iteration 1.

If so, why not call `xm.add_step_closure` before `xm.optimizer_step(optimizer)` in the loop rather than after?   Although I see it would be equivalent, it is confusing since I believe the closure would depend on the mark_step from the loader being run, rather than from the `optimizer_step` call.

**Comment by JackCaoG** [comment]
`optimizer_step` does not call `mark_step` unless you set the barrier to true. In above example I believe `mark_step` will be triggered by the parallelLoader.

**Comment by JackCaoG** [comment]
check https://github.com/pytorch/xla/blob/7a6aa398b0c030b008b035b21b3f6d9e4c6b81d1/torch_xla/core/xla_model.py#L879

**Comment by hrbigelow** [comment]
Ahh yes, I forgot about the barrier flag.  It looks like either way the first step closure will be run during the second iteration of the loop, but it doesn't matter that much.  Thanks for clarifying.
